src/python/pants/backend/python/util_rules/pex_cli.py

```diff
@@ -120,6 +127,73 @@ def __post_init__(self) -> None:
             raise ValueError("`--pex-root` flag not allowed. We set its value for you.")
 
 
+@union
+@dataclass(frozen=True)
+class PexKeyringConfigurationRequest:
+    """Request sent to keyring plugins to request credentials to expose to Pex/Pip."""
+
+    name: ClassVar[str]
+
+
+@dataclass(frozen=True)
+class PexKeyringConfigurationResponse:
+    """Response containing credentials to expose to Pex/Pip via simulating the `keyring` package's
+    binary."""
+
+    # Nested map from SITE -> (USER, PASSWORD).
+    credentials: FrozenDict[str, tuple[str, str]] | None
+
+
+# The following script is injected into any Pex CLI process where credentials will be exposed via
+# the `keyring` package. Pip will be configured to use its `subprocess` mode which will obtain
+# credentials by invoking this "keyring" binary from the PATH.
+#
+# Credentials are stored in the file stored in the `__PANTS_KEYRING_DATA` environment variable.
```

Comment:
I would emphasize (via name and comment) that the value in the env var is a path, not actual data content.